ui/core/src/model/variables.ts

```diff
@@ -13,29 +13,54 @@
 
 import { Definition, JsonObject } from './definitions';
 
-/**
- * Variable definition options that are common to all variables.
- */
-export interface VariableDefinition<Options extends JsonObject = JsonObject> extends Definition<Options> {
-  display: VariableDisplayOptions;
-  selection: VariableSelectionOptions;
-  capturing_regexp?: string;
+export const DEFAULT_ALL_VALUE = '$__all' as const;
+
+export type VariableName = string;
+
+export type VariableState = {
+  value: VariableValue;
+  options?: VariableOption[];
+  loading: boolean;
+  error?: Error | null;
+};
+
+export type VariableOption = { label: string; value: string };
+
+export type VariablesState = Record<VariableName, VariableState>;
+
+export type VariableValue = string | string[] | null;
+
+export interface IVariable<TKind extends string, TOptions extends JsonObject = JsonObject>
+  extends Definition<TOptions> {
+  kind: TKind;
+  name: VariableName;
+  display?: {
+    label?: string;
+    hidden?: boolean;
+  };
+  defaultValue?: VariableValue;
+  options: TOptions;
 }
 
-export interface VariableDisplayOptions extends JsonObject {
-  hide?: boolean;
-  label: string;
+export interface TextVariableOptions extends JsonObject {
+  value: string;
 }
 
-export type VariableSelectionOptions = SingleSelectOptions | MultiSelectOptions;
+export type TextVariableDefintion = IVariable<'TextVariable', TextVariableOptions>;
 
-export type SingleSelectOptions = {
-  default_value: string;
+export type ListVariableOptions<TKind extends string, TOptions extends JsonObject> = {
+  allowMultiple?: boolean;
+  allowAllValue?: boolean;
+  customAllValue?: string;
+  kind: TKind;
+  options: TOptions;
```

Comment:
Yeah I'm open to suggestion on what definition is cleaner (and more consistent) here. 

So instead of this:
```json
{
  "kind": "ListVariable",
  "name": "MyVariable",
  "options": {
    "allowMultiple": true,
    "kind": "PromQLLabelValues",
    "options": {
      "label": "job"
    }
  }
}
```

Do this?
```json

{
  "kind": "ListVariable",
  "name": "MyVariable",
  "options": {
    "allowMultiple": true,
    "optionsLoader": {
      "kind": "PromQLLabelValues",
      "options": {
        "label": "job"
      }
    }
  }
}
```

Reply:
Yeah, basically that. It just felt weird to me to have `options.kind` and `options.options` 😂

Reply:
well actually I would rather prefer to go with the first solution since it is consistent with the Panels, the datasource and basically the whole rest of object definition.

`kind` always defined the type of `options` => aka `spec` ;)

Reply:
beside the first solution is flatten than the second one. So less painful to write if you write the JSON by yourself ;)

Reply:
Anything common to the ListVariable should go on the first `options`. The second level is conditioned by the `kind`.
Since `kind` is common to the `ListVariable` definition, it should go in the first `options` level